consensus/view_change_construct.go

```diff
@@ -0,0 +1,422 @@
+
+	sig3arr := []*bls_core.Sign{}
+	for _, sig := range vc.viewIDSigs[viewID] {
+		sig3arr = append(sig3arr, sig)
+	}
+	// even we check here for safty, m3 type signatures must >= 2f+1
+	if len(sig3arr) > 0 {
+		m3Sig := bls_cosi.AggregateSig(sig3arr)
+		vc.getLogger().Info().Int("len", len(sig3arr)).Msg("[GetM3Bitmap] M3 (ViewID) type signatures")
+		return m3Sig.Serialize(), vc.viewIDBitmap[viewID].Bitmap
+	}
+	vc.getLogger().Info().Uint64("viewID", viewID).Msg("[GetM3Bitmap] No M3 (ViewID) type signatures")
+	return nil, nil
+}
+
+// VerifyNewViewMsg returns true if the new view message is valid
+func (vc *viewChange) VerifyNewViewMsg(recvMsg *FBFTMessage) (*types.Block, error) {
+	vc.vcLock.Lock()
+	defer vc.vcLock.Unlock()
+
+	if recvMsg.M3AggSig == nil || recvMsg.M3Bitmap == nil {
+		return nil, errors.New("[VerifyNewViewMsg] M3AggSig or M3Bitmap is nil")
+	}
+
+	senderKey := recvMsg.SenderPubkey
+	senderKeyStr := senderKey.Bytes.Hex()
+
+	// first time received the new view message
+	_, ok := vc.newViewMsg[recvMsg.ViewID]
+	if !ok {
+		newViewMap := map[string]uint64{}
+		vc.newViewMsg[recvMsg.ViewID] = newViewMap
+	}
+
+	_, ok = vc.newViewMsg[recvMsg.ViewID][senderKeyStr]
+	if ok {
+		return nil, errors.New("[VerifyNewViewMsg] received new view message already")
+	}
+
+	vc.newViewMsg[recvMsg.ViewID][senderKeyStr] = recvMsg.BlockNum
+
+	m3Sig := recvMsg.M3AggSig
+	m3Mask := recvMsg.M3Bitmap
+
+	viewIDBytes := make([]byte, 8)
+	binary.LittleEndian.PutUint64(viewIDBytes, recvMsg.ViewID)
+
+	if !m3Sig.VerifyHash(m3Mask.AggregatePublic, viewIDBytes) {
+		return nil, errors.New("[VerifyNewViewMsg] Unable to Verify Aggregated Signature of M3 (ViewID) payload")
+	}
+
+	m2Mask := recvMsg.M2Bitmap
+	if recvMsg.M2AggSig != nil {
+		m2Sig := recvMsg.M2AggSig
+		if !m2Sig.VerifyHash(m2Mask.AggregatePublic, NIL) {
+			return nil, errors.New("[VerifyNewViewMsg] Unable to Verify Aggregated Signature of M2 (NIL) payload")
+		}
+	}
+
+	// check when M3 sigs > M2 sigs, then M1 (recvMsg.Payload) should not be empty
+	preparedBlock := &types.Block{}
+	if len(recvMsg.Payload) >= ValidPayloadLength && len(recvMsg.Block) != 0 {
+		if err := rlp.DecodeBytes(recvMsg.Block, preparedBlock); err != nil {
+			return nil, errors.New("[VerifyNewViewMsg] Unparseable prepared block data")
+		}
+
+		blockHash := recvMsg.Payload[:32]
+		preparedBlockHash := preparedBlock.Hash()
+		if !bytes.Equal(preparedBlockHash[:], blockHash) {
+			return nil, errors.New("[VerifyNewViewMsg] Prepared block hash doesn't match msg block hash")
+		}
+		if err := vc.blockVerifier(preparedBlock); err != nil {
+			return nil, errors.New("[VerifyNewViewMsg] Prepared block verification failed")
+		}
+		return preparedBlock, nil
+	}
+	return nil, nil
+}
+
+// VerifyViewChangeMsg returns the type of the view change message after verification
+func (vc *viewChange) VerifyViewChangeMsg(recvMsg *FBFTMessage, members multibls.PublicKeys) (ViewChangeMsgType, *types.Block, error) {
```

Comment:
This function is more like "ProcessViewChangeMsg" then just verify, because it's modifying the state (not a pure function)

Reply:
will rename